bigquery/snippets/managedwriter/bigquerystorage_append_rows_default.go

```diff
@@ -0,0 +1,199 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package managedwriter
+
+// [START bigquerystorage_write_default_complexschema]
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"time"
+
+	"cloud.google.com/go/bigquery/storage/managedwriter"
+	"cloud.google.com/go/bigquery/storage/managedwriter/adapt"
+	"github.com/GoogleCloudPlatform/golang-samples/bigquery/snippets/managedwriter/exampleproto"
+	"google.golang.org/protobuf/proto"
+)
+
+// generateExampleMessages generates a slice of serialized protobuf messages using a statically defined
+// and compiled protocol buffer file, and returns the binary serialized representation.
+func generateExampleDefaultMessages(numMessages int) ([][]byte, error) {
+	msgs := make([][]byte, numMessages)
+	for i := 0; i < numMessages; i++ {
+
+		random := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+		// Our example data embeds an array of structs, so we'll construct that first.
+		sList := make([]*exampleproto.SampleStruct, 5)
+		for i := 0; i < int(random.Int63n(5)+1); i++ {
+			sList[i] = &exampleproto.SampleStruct{
+				SubIntCol: proto.Int64(random.Int63()),
+			}
+		}
+
+		m := &exampleproto.SampleData{
+			BoolCol:    proto.Bool(true),
+			BytesCol:   []byte("some bytes"),
+			Float64Col: proto.Float64(3.14),
+			Int64Col:   proto.Int64(123),
+			StringCol:  proto.String("example string value"),
+
+			// These types require special encoding/formatting to transmit.
+
+			// DATE values are number of days since the Unix epoch.
+
+			DateCol: proto.Int32(int32(time.Now().UnixNano() / 86400000000000)),
+
+			// DATETIME uses the literal format.
+			DatetimeCol: proto.String("2022-01-01 12:13:14.000000"),
+
+			// GEOGRAPHY uses Well-Known-Text (WKT) format.
+			GeographyCol: proto.String("POINT(-122.350220 47.649154)"),
+
+			// NUMERIC and BIGNUMERIC can be passed as string, or more efficiently
+			// using a packed byte representation.
+			NumericCol:    proto.String("99999999999999999999999999999.999999999"),
+			BignumericCol: proto.String("578960446186580977117854925043439539266.34992332820282019728792003956564819967"),
+
+			// TIME also uses literal format.
+			TimeCol: proto.String("12:13:14.000000"),
+
+			// TIMESTAMP uses microseconds since Unix epoch.
+			TimestampCol: proto.Int64(time.Now().UnixNano() / 1000),
+
+			// Int64List is an array of INT64 types.
+			Int64List: []int64{2, 4, 6, 8},
+
+			// This is a required field, and thus must be present.
```

Comment:
question: Is this required by BigQuery?

Reply:
It's a required field of the schema (aka nulls not allowed).

Reply:
augmenting the comment to try to make that clearer.